mlir/lib/QEC/Transforms/AbsorbCliffordToPPM.cpp

```diff
@@ -0,0 +1,215 @@
+    SmallVector<StringRef> pauliProductArrayRef = removeIdentityPauli(rhs, newRHSOperands);
+    mlir::ArrayAttr pauliProduct = rewriter.getStrArrayAttr(pauliProductArrayRef);
+
+    // Get the type list from new RHS
+    SmallVector<Type> newOutQubitTypes;
+    for (auto qubit : newRHSOperands) {
+        newOutQubitTypes.push_back(qubit.getType());
+    }
+
+    Type mresType = rhs.getMres().getType();
+
+    auto newPPM =
+        rewriter.create<PPMeasurementOp>(rhs->getLoc(), mresType, newOutQubitTypes, pauliProduct,
+                                         rhs.getRotationSign(), newRHSOperands);
+    rewriter.moveOpBefore(newPPM, rhs);
+
+    // Update the use of value in newRHSOperands
+    for (unsigned i = 0; i < newRHSOperands.size(); i++) {
+        newRHSOperands[i].replaceAllUsesExcept(newPPM.getOutQubits()[i], newPPM);
+    }
+
+    rewriter.replaceAllUsesWith(rhs.getMres(), newPPM.getMres());
+    rewriter.replaceAllUsesWith(rhs.getOutQubits(), rhs.getInQubits());
+
+    rewriter.eraseOp(rhs);
+
+    sortTopologically(lhs->getBlock());
+}
+
+bool shouldRemovePPR(PPRotationOp op)
+{
+    if (op->getUsers().empty())
+        return true;
+
+    SetVector<Operation *> slice;
+    getForwardSlice(op, &slice);
+
+    for (auto op : slice) {
+
+        if (isa<PPMeasurementOp>(op))
+            return false;
+
+        if (isa<catalyst::quantum::DeallocOp>(op))
+            return true;
+    }
+
+    return false;
+}
+
+struct AbsorbCliffordToPPM : public OpRewritePattern<PPMeasurementOp> {
+    using OpRewritePattern::OpRewritePattern;
+
+    LogicalResult matchAndRewrite(PPMeasurementOp op, PatternRewriter &rewriter) const override
+    {
+        return visitPPMeasurementOp(op, [&](PPRotationOp pprOp) {
+            PauliWordPair normOps = normalizePPROps(pprOp, op);
+
+            if (normOps.first.commutes(normOps.second)) {
+                moveCliffordPastPPM(normOps.first, normOps.second, nullptr, rewriter);
+            }
+            else {
+                auto resultStr = normOps.first.computeCommutationRulesWith(normOps.second);
+                moveCliffordPastPPM(normOps.first, normOps.second, &resultStr, rewriter);
+            }
+            return success();
+        });
+    }
+};
+
+struct RemoveDeadPPR : public OpRewritePattern<PPRotationOp> {
+    using OpRewritePattern::OpRewritePattern;
+
+    LogicalResult matchAndRewrite(PPRotationOp op, PatternRewriter &rewriter) const override
+    {
+        if (shouldRemovePPR(op)) {
+            rewriter.replaceOp(op, op.getInQubits());
+            return success();
+        }
+        return failure();
+    }
+};
+
+} // namespace
+
+namespace catalyst {
+namespace qec {
+
+void populateAbsorbCliffordToPPMPatterns(mlir::RewritePatternSet &patterns)
+{
+    patterns.add<AbsorbCliffordToPPM>(patterns.getContext(), 1);
+}
+
+void populateRemoveDeadPPRPatterns(mlir::RewritePatternSet &patterns)
+{
+    patterns.add<RemoveDeadPPR>(patterns.getContext(), 1);
+}
```

Comment:
I think the point of the `populate*` functions (which are fully optional), is that many patterns can be regrouped together and added to a pattern set in one step, otherwise they are kind of useless (for a single pattern) 😅 
In this case, the `RemoveDeadPPR` and `CommuteCliffordPastPPM` patterns together form the "absorption" transformation, so it makes sense to group them together. It would also be good to rename the `AbsorbCliffordToPPM` pattern to `CommuteCliffordPastPPM` since that is a more accurate description.
```suggestion
void populateAbsorbCliffordToPPMPatterns(mlir::RewritePatternSet &patterns)
{
    patterns.add<CommuteCliffordPastPPM>(patterns.getContext(), 1);
    patterns.add<RemoveDeadPPR>(patterns.getContext(), 1);
}
```

Reply:
Yes, it makes sense. Actually, I thought we could reuse `RemoveDeadPPR` somewhere else, but if doing so, it should be separated into another file. I agree with group them together.